services-custom/s3-transfer-manager/src/main/java/software/amazon/awssdk/transfer/s3/internal/utils/ResumableFileDownloadSerializer.java

```diff
@@ -0,0 +1,269 @@
+import software.amazon.awssdk.protocols.json.internal.unmarshall.JsonUnmarshallerRegistry;
+import software.amazon.awssdk.protocols.json.internal.unmarshall.SimpleTypeJsonUnmarshaller;
+import software.amazon.awssdk.protocols.jsoncore.JsonNode;
+import software.amazon.awssdk.protocols.jsoncore.JsonNodeParser;
+import software.amazon.awssdk.services.s3.model.GetObjectRequest;
+import software.amazon.awssdk.transfer.s3.DownloadFileRequest;
+import software.amazon.awssdk.transfer.s3.ResumableFileDownload;
+import software.amazon.awssdk.transfer.s3.S3TransferManager;
+import software.amazon.awssdk.utils.Logger;
+
+@SdkInternalApi
+public final class ResumableFileDownloadSerializer {
+    private static final Logger log = Logger.loggerFor(S3TransferManager.class);
+
+    private ResumableFileDownloadSerializer() {
+    }
+
+    /**
+     * Serializes an instance of {@link ResumableFileDownload} into valid JSON. This object contains a nested GetObjectRequest and
+     * therefore makes use of the standard JSON marshalling classes.
+     */
+    public static byte[] toJson(ResumableFileDownload resumableFileDownload) {
+        StructuredJsonGenerator generator = jsonGenerator();
+        JsonMarshallerRegistry marshallerRegistry = createMarshallerRegistry();
+        JsonMarshallerContext context = jsonMarshallerContext(generator, marshallerRegistry);
+
+        generator.writeStartObject();
+        SimpleTypeJsonMarshaller.LONG.marshall(resumableFileDownload.bytesTransferred(), context, "bytesTransferred", null);
+        SimpleTypeJsonMarshaller.INSTANT.marshall(resumableFileDownload.fileLastModified(), context, "fileLastModified", null);
+        if (resumableFileDownload.totalSizeInBytes().isPresent()) {
+            SimpleTypeJsonMarshaller.LONG.marshall(resumableFileDownload.totalSizeInBytes().get(),
+                                                   context,
+                                                   "totalSizeInBytes",
+                                                   null);
+        }
+        if (resumableFileDownload.s3ObjectLastModified().isPresent()) {
+            SimpleTypeJsonMarshaller.INSTANT.marshall(resumableFileDownload.s3ObjectLastModified().get(),
+                                                      context,
+                                                      "s3ObjectLastModified",
+                                                      null);
+        }
+        marshallDownloadFileRequest(resumableFileDownload.downloadFileRequest(), context);
+        generator.writeEndObject();
+
+        return generator.getBytes();
+    }
+
+    /**
+     * At this point we do not need to persist the TransferRequestOverrideConfiguration, because it only contains listeners and
+     * they are not used in the resume operation.
+     */
+    private static void marshallDownloadFileRequest(DownloadFileRequest fileRequest, JsonMarshallerContext context) {
+        context.jsonGenerator().writeFieldName("downloadFileRequest");
```

Comment:
nit: assign `jsonGenerator` to a local so the code is a bit easier to read. Ditto in other places

Reply:
will do